Feat/ecoinvent interface

will7200 opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and
review the checklist below.

Contributors guide: ./CONTRIBUTING.md
-->

- Integrates `ecoinvent_interface` package instead of `eidl`.
- Uses a versioned biosphere database instead of the default batabase named `biosphere`

Closes #7 


## Checklist
<!--
Remove items that do not apply. For completed items, change [ ] to [x] or you can click the checkboxes once your 
pull-request is published.
-->

- [x] Keep pull requests small so they can be easily reviewed.
- [ ] Update the documentation, please follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] Update tests.
- [ ] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `feature`, `ui`, `change`, `documentation`, `breaking`, `ci`
      as they show up in the changelog.
- [ ] Link this PR to related issues by using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
- [ ] Add a milestone to the PR for the intended release.
- [ ] Request a review from another developer.

<!--
NOTE: these things are not required to open a PR and can be done afterwards,
while the PR is open.
-->

will7200 committed:
feat: support login workflow with ecoinvent_interface

will7200 committed:
feat: update ecoinvent version page

will7200 committed:
feat: implement downloading and extract file

also cleanup the login signal and move them out of the downloader

will7200 committed:
fix: import errors

will7200 committed:
fix: always download ecospold file

will7200 committed:
fix: activity browser sqlite3 superclass

will7200 committed:
feat: support listing cached downloads

cmutel reviewed (changes requested):
Nice, I am impressed that you saw some basic patterns have changed.

In the future I am pretty sure we can remove `ABEcoinventDownloader` completely.

One last change needed (building on https://github.com/LCA-ActivityBrowser/activity-browser/issues/1301): The current code assumes that a database called "biosphere3" is present, but this is an antipattern. Instead, we should set the default biosphere preference to the biosphere database corresponding to the ecoinvent import. So, if we imported ecoinvent 3.9.1, we would end up with two databases: `ecoinvent-3.9.1-something` and `ecoinvent-3.9.1-biosphere`. We need to run the following:

```python
import bw2data as bd
bd.preferences['biosphere_database'] = 'ecoinvent-3.9.1-biosphere'
bd.preferences.flush()
```

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -823,17 +831,23 @@ def run_forwast(self) -> None:
+        return filepath
 
-    def run_extract(self, temp_dir: Path) -> None:
+    def run_extract(self, archive_file: Path, temp_dir: Path) -> None:
```
I kept extracting local to AB for the following reasons:
1. AB subclasses its own Extractor to keep the amount of files extract for UI progress
2. AB supports extracting from a local file, so a implementation is still required to support this
3. `ecoinvent_interface` has inconsistent behavior when downloading the same file when `extract=False` vs `extract=True`

cmutel replied:
> AB subclasses its own Extractor

Two wrongs don't make a right 😛 But we can leave it as is for now.

> ecoinvent_interface has inconsistent behavior when downloading the same file when extract=False vs extract=True

I hope so, those are very different actions? Can you clarify?

will7200 replied:
If the file in question doesn't exist previously than extract true vs false works as expected.
I meant when a previously downloaded archive was downloaded, switching between extract true vs false, will give you the initial value.
If not extracted the first time, and then wanting to extract the second time, you will get the tar link regardless. And vice versa
```python
>>> import ecoinvent_interface as ei
>>> r = ei.EcoinventRelease(ei.Settings())
>>> r.get_release("3.8", "cutoff", ei.ReleaseType.ecospold, extract=False)
PosixPath('/Users/wflores/Library/Application Support/EcoinventInterface/cache/ecoinvent 3.8_cutoff_ecoSpold02.7z')
>>> r.get_release("3.8", "cutoff", ei.ReleaseType.ecospold, extract=True)
PosixPath('/Users/wflores/Library/Application Support/EcoinventInterface/cache/ecoinvent 3.8_cutoff_ecoSpold02.7z')
```

will7200 replied:
I am all for changing the behavior, just let me know the direction and i can execute. Just trying to retain the existing feature set.

cmutel replied:
It's fine for now, let's get it working before polishing to perfection. Thanks.

will7200 replied:
changed to extract by default, but double check when extracted if the passed in is a directory(has been extracted) or file.

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -970,11 +985,20 @@ def __init__(self, parent=None):
-        self.password_edit.setPlaceholderText("ecoinvent password"),
         self.password_edit.setEchoMode(QtWidgets.QLineEdit.Password)
+        if eco_settings.password:
+            self.password_edit.setText(eco_settings.password)
```
Password is shown as a mask to the user, and they can't copy the value, but they can modify.

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -970,11 +985,20 @@ def __init__(self, parent=None):
         super().__init__(parent)
         self.wizard = parent
         self.complete = False
+        eco_settings = ecoinvent_interface.Settings()
```
Sure we can do that

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -344,13 +356,13 @@ def __init__(self, parent=None):
 
     def initializePage(self):
-        self.stored_dbs = eidl.eidlstorage.stored_dbs
+        self.stored_dbs = ecoinvent_interface.CachedStorage()
```
Yes you are correct, there are different formats available and releases available. I filter explicitly for Ecospold data

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1316,27 +1364,115 @@ class ImportSignals(QtCore.QObject):
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.extraction_process = None
+class ABEcoinventDownloader(object):
```
It's habitual for me to do that, will update

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -344,7 +344,10 @@ def __init__(self, parent=None):
 
     def initializePage(self):
-        self.stored_dbs = eidl.eidlstorage.stored_dbs
+        # TODO: get this from eco_invent
```
We can just squash merge this PR. I am getting acclimated to this repo, I wanted to chunk my work instead of doing it all at once.

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1039,7 +1042,7 @@ def nextId(self):
 
 class LoginThread(QtCore.QThread):
-    def __init__(self, downloader, parent=None):
+    def __init__(self, downloader: "ABEcoinventDownloader", parent=None):
```
type hints in quotes are just as valid as ones without. They signal a [Forward Reference](https://www.python.org/dev/peps/pep-0484/#forward-references). mypy follows the pep spec for type hinting.
give it a shot. This entire wizard uses references that are defined afterwards, so in an effort to keep things minimal left as is.

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1048,7 +1051,12 @@ def update(self, username: str, password: str) -> None:
-        self.downloader.login()
+        try:
+            self.downloader.login()
+        except Exception as e:
```
Notice that in this particular instance, it is being run in a separate thread, if a broad exception is not used, an error may occur that will never be signaled back to the user.

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1316,14 +1324,50 @@ class ImportSignals(QtCore.QObject):
-        super().__init__(*args, **kwargs)
+# TODO: reimplement downloader using ecoinvent_interface
+class ABEcoinventDownloader(object):
+    def __init__(self, version=None, system_model=None):
```
added

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1316,14 +1324,50 @@ class ImportSignals(QtCore.QObject):
+        return self._settings.username
+
+    @username.setter
+    def username(self, value):
```
added

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1316,14 +1324,50 @@ class ImportSignals(QtCore.QObject):
+        return self._settings.password
+
+    @password.setter
+    def password(self, value):
```
added

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1316,14 +1324,50 @@ class ImportSignals(QtCore.QObject):
+
+        import_signals.login_success.emit(login_success)
+
+    # TODO: don't think we need this anymore
```
squashing this PR since the end result doesn't have them

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1316,14 +1324,50 @@ class ImportSignals(QtCore.QObject):
+        import_signals.login_success.emit(login_success)
+
+    # TODO: don't think we need this anymore
     def extract(self, target_dir):
```
later commits implement them

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1062,7 +1063,7 @@ def run(self):
     def __init__(self, parent=None):
         super().__init__(parent)
-        self.wizard = self.parent()
+        self.wizard: "DatabaseImportWizard" = self.parent()
```
In this context, looking at it, the most I can assume is `wizard: any | None`.
Just because there is one usage, doesn't mean that it can be inferred that is indeed `DatabaseImportWizard`.
My 2 cents, if it is bothersome i can remove it.

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -744,7 +752,7 @@ def report_failed_unarchive(self, file: str) -> None:
     def __init__(self, downloader, parent=None):
         super().__init__(parent)
-        self.downloader = downloader
+        self.downloader: "ABEcoinventDownloader" = downloader
```
Looks like any to me, moving this up to the parameter.
If not desired i can remove it

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -880,14 +881,14 @@ def run_import(self, import_dir: Path) -> None:
             )
             importer.apply_strategies()
-            importer.write_database(backend="activitybrowser")
+            importer.write_database(backend="sqlite")
```
There is two options `sqlite` and `iotable`, activitybrowser wraps sqlite for UI progress bars.

@cmutel 

Do we want to support `iotable`?

cmutel replied:
No, this is an internal implementation detail which users don't need to know about. This is fine.

will7200 commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -1318,22 +1321,28 @@ def extract(cls, dirpath: str, db_name: str, *args, **kwargs):
 
 
 bd.config.backends["activitybrowser"] = ActivityBrowserBackend
+# config is no longer enough to provide an additional backend
```
bw2data doesn't have an api to register a new backend, i think this necessitates a call out since it may break more easily.

cmutel replied:
> bw2data doesn't have an api to register a new backend

[Respectfully disagree](https://github.com/brightway-lca/brightway2-data/blob/main/bw2data/subclass_mapping.py#L5)

will7200 replied:
IMO a undocumented dictionary variable does not constitute as a stable contract, as opposed to a named function whose lifecycle can be managed (docs, depreations, etc)

I use DATABASE_BACKEND_MAPPING, but i called it out

will7200 committed:
chore: update typings

will7200 committed:
chore: remove explict inheritance for downloader

cmutel commented on activity_browser/ui/wizards/db_import_wizard.py:
```diff
@@ -881,7 +883,8 @@ def run_import(self, import_dir: Path) -> None:
             )
             importer.apply_strategies()
-            importer.write_database(backend="sqlite")
+            # backend is a custom implementation that wraps sqlite database
```
Yeah, the `activitybrowser` backend should go away, but one step at a time. And comments are allowed in general :)

will7200 committed:
feat: skip login page if user has valid credentials

will7200 committed:
fix: handle case when version is empty

going back and forth between the wizard pages can set this to an empty string and cause an exception

will7200 committed:
fix: filter catalog items for ecoSpold02 files only

will7200 committed:
feat: update biosphere_database preference on import

will7200 commented:
Added the bd.preferences to update on eco version page.
Do we need to run UpdateBiosphere?

cmutel commented:
> Do we need to run UpdateBiosphere?

No. In general this isn't a great idea, and should require manual action by the user after the import complete.

will7200 committed:
fix: handle case when no ecoinvent credentials are pre-configured

will7200 committed:
refactor: use extract of ecoinvent

still have a fallback if the provided file is 7z

will7200 committed:
feat: add page to create versioned biosphere database

will7200 commented:
I ended having to integrate UpdateBiosphere for the newly versioned database created. Otherwise the end user will go through the flow of trying to import the database, only to get the `update the biosphere by using 'File' -> 'Update biosphere...'` message, having to go through the wizard twice to get it to work properly.

This versioned database page only gets shown to the end user if that particular database does not exist already.
